cmd/git-sync/main.go

```diff
@@ -569,66 +563,108 @@ func revIsHash(ctx context.Context, rev, gitRoot string) (bool, error) {
-			return false, "", err
-		}
-		if local == remote {
-			log.V(1).Info("no update required", "rev", rev, "local", local, "remote", remote)
-			return false, "", nil
+		// Make sure the directory we found is actually usable.
+		if err := sanityCheck(ctx, gitRoot, dest); err != nil {
+			// Maybe a previous run crashed?  Git won't use this dir.
+			log.V(0).Info("git root exists but failed sanity checks, cleaning up", "path", gitRoot)
+			if err := os.RemoveAll(gitRoot); err != nil {
+				return false, "", err
+			}
+			// Clone it and get the hash.
+			hash, err := cloneAndHash(ctx, repo, rev, depth, gitRoot)
+			if err != nil {
+				return false, "", err
+			}
+			hashToSync = hash
+		} else {
+			// Not the first time. Figure out if the rev has changed.
+			local, err := localHashForRev(ctx, rev, gitRoot)
+			if err != nil {
+				return false, "", err
+			}
+			remote, err := remoteHashForRev(ctx, rev, gitRoot)
+			if err != nil {
+				return false, "", err
+			}
+			if local == remote {
+				log.V(1).Info("no update required", "rev", rev, "local", local, "remote", remote)
+				return false, local, nil
+			}
+			log.V(0).Info("update required", "rev", rev, "local", local, "remote", remote)
+
+			if err := syncFromRemote(ctx, gitRoot, depth); err != nil {
+				return true, remote, err
+			}
+			hashToSync = remote
 		}
-		log.V(0).Info("update required", "rev", rev, "local", local, "remote", remote)
-		hash = remote
 	}
 
-	return true, hash, addWorktreeAndSwap(ctx, gitRoot, dest, branch, rev, depth, hash)
+	return true, hashToSync, addWorktreeAndPublish(ctx, gitRoot, dest, depth, hashToSync)
 }
 
-// getRevs returns the local and upstream hashes for rev.
-func getRevs(ctx context.Context, localDir, branch, rev string) (string, string, error) {
-	// Ask git what the exact hash is for rev.
-	local, err := localHashForRev(ctx, rev, localDir)
+func cloneAndHash(ctx context.Context, repo, rev string, depth int, gitRoot string) (string, error) {
+	if err := cloneRepo(ctx, repo, depth, gitRoot); err != nil {
+		return "", err
+	}
+	hash, err := localHashForRev(ctx, rev, gitRoot)
 	if err != nil {
-		return "", "", err
+		return "", err
 	}
+	return hash, nil
+}
 
-	// Build a ref string, depending on whether the user asked to track HEAD or a tag.
-	ref := ""
-	if rev == "HEAD" {
-		ref = "refs/heads/" + branch
-	} else {
-		ref = "refs/tags/" + rev
+func syncFromRemote(ctx context.Context, gitRoot string, depth int) error {
+	log.V(0).Info("fetching from remote")
+
+	// Update from the remote.
+	args := []string{"fetch", "--prune"}
+	if depth != 0 {
+		args = append(args, "--depth", strconv.Itoa(depth))
+	}
+	if _, err := runCommand(ctx, gitRoot, *flGitCmd, args...); err != nil {
+		return err
 	}
 
-	// Figure out what hash the remote resolves ref to.
-	remote, err := remoteHashForRef(ctx, ref, localDir)
-	if err != nil {
-		return "", "", err
+	// GC the repo.
+	if _, err := runCommand(ctx, gitRoot, *flGitCmd, "gc", "--prune=all"); err != nil {
```

Comment:
```suggestion
	if _, err := runCommand(ctx, gitRoot, *flGitCmd, "gc", "--prune=now"); err != nil {
```

Reply:
What are you trying to clean up with gc? Stale worktrees? Wouldn't that need a "worktree remove" beforehand or else it never becomes stale?